components/script/dom/notification.rs

```diff
@@ -0,0 +1,545 @@
+        self.image.clone().unwrap_or_default()
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-icon>
+    fn Icon(&self) -> USVString {
+        // step 1: If there is no this’s notification’s icon URL, then return the empty string.
+        // step 2: Return this’s notification’s icon URL, serialized.
+        self.icon.clone().unwrap_or_default()
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-badge>
+    fn Badge(&self) -> USVString {
+        // step 1: If there is no this’s notification’s badge URL, then return the empty string.
+        // step 2: Return this’s notification’s badge URL, serialized.
+        self.badge.clone().unwrap_or_default()
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-renotify>
+    fn Renotify(&self) -> bool {
+        self.renotify
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-silent>
+    fn GetSilent(&self) -> Option<bool> {
+        self.silent
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-requireinteraction>
+    fn RequireInteraction(&self) -> bool {
+        self.require_interaction
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-data>
+    #[allow(unsafe_code)]
+    fn Data(&self, _cx: SafeJSContext, mut retval: MutableHandleValue) {
+        retval.set(self.data.get());
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-actions>
+    fn Actions(&self, cx: SafeJSContext, retval: MutableHandleValue) {
+        // step 1: Let frozenActions be an empty list of type NotificationAction.
+        let mut frozen_actions: Vec<NotificationAction> = Vec::new();
+
+        // step 2: For each entry of this’s notification’s actions
+        for action in self.actions.iter() {
+            let action = NotificationAction {
+                action: action.name.clone(),
+                title: action.title.clone(),
+                // If entry’s icon URL is non-null,
+                // then set action["icon"] to entry’s icon URL, icon_url, serialized.
+                icon: action.icon_url.clone(),
+            };
+
+            // TODO: step 2.5: Call Object.freeze on action, to prevent accidental mutation by scripts.
+            // step 2.6: Append action to frozenActions.
+            frozen_actions.push(action);
+        }
+
+        // step 3: Return the result of create a frozen array from frozenActions.
+        to_frozen_array(frozen_actions.as_slice(), cx, retval);
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-vibrate>
+    fn Vibrate(&self, cx: SafeJSContext, retval: MutableHandleValue) {
+        to_frozen_array(self.vibration_pattern.as_slice(), cx, retval);
+    }
+    /// <https://notifications.spec.whatwg.org/#dom-notification-timestamp>
+    fn Timestamp(&self) -> u64 {
+        self.timestamp
+    }
+    /// <https://notifications.spec.whatwg.org/#handle-close-events>
```

Comment:
```suggestion
    /// <https://notifications.spec.whatwg.org/#dom-notification-close>
```
I think the implementation here belongs in a method on GlobalScope instead, since that's where the list of notifications will live.